### ISSUE TITLE
Font Color in bast script

### ISSUE THREAD

**Issue by Zorgodon** [felixhageloh/uebersicht#136]
Any sort of font coloring in bash scripts is not interpreted+displayed by ubersicht.

**Comment by soberstadt** [comment]
Could you give me some more information? How is the color getting set?

Currently, all styles are defined by css. As this is a WebView rendering all of the content.

**Comment by Zorgodon** [comment]
Right, I run a bash script to give me today's date in the format "Thursday 05 November" , with the "05" two numbers coloured red and the rest of the date white. I used to be able to achieve this in Geektool. The script uses colouring in the printf command. Ubersicht, instead of colouring the text, outputs the printf attributes directly.
The script is:

#!/bin/bash
Reset="\e[0m"
ESC="\e["
RES="0"
RedFont="31"
BGColor="$RES"
Face="$RES"
FontColor="$RES"
dateDuJour=$(date "+%d")
dateDuJourTS=$(date -j $dateDuJour "+%s")
dateJourJ=$(date -r $dateDuJourTS "+%A %d %B")
dateJourJJour=$(echo $dateJourJ | awk '{print $1}')
dateJourJNum=$(echo $dateJourJ | awk '{print $2}')
dateJourJMois=$(echo $dateJourJ | awk '{print $3}')
printf "%8.9s ${ESC}${RedFont}m%s$Reset %s\n" $dateJourJJour $dateJourJNum $dateJourJMois

Übersicht outputs: Thursday [31m05[0m November (instead of colouring the "05", it literally prints the colour attributes!!!)
I am probably doing something wrong, I am trying to use this script, not written by me, originally for Geektool in Ubersicht. Maybe you could suggest an alternative way to colour the "05" in the date?

The full widget:

command: """
#!/bin/bash
Reset="\e[0m"
ESC="\e["
RES="0"
RedFont="31"
BGColor="$RES"
Face="$RES"
FontColor="$RES"
dateDuJour=$(date "+%d")
dateDuJourTS=$(date -j $dateDuJour "+%s")
dateJourJ=$(date -r $dateDuJourTS "+%A %d %B")
dateJourJJour=$(echo $dateJourJ | awk '{print $1}')
dateJourJNum=$(echo $dateJourJ | awk '{print $2}')
dateJourJMois=$(echo $dateJourJ | awk '{print $3}')
printf "%8.9s ${ESC}${RedFont}m%s$Reset %s\n" $dateJourJJour $dateJourJNum $dateJourJMois
"""

refreshFrequency: 500

style: """
  top: 200px
  right: 0px
  color: #FFFFFF

  .output
    font-family: Roboto-Italic
    font-size: 60px
    font-weight: 100
    text-shadow: 0 0px 0px #000000;
"""

render: (output) -> """
  <div class="output">#{output}</div>
"""

Thx. Only starting to get my head around coffeescript!
-- 
Ivan Bogush
Sent with Airmail

On 5 November 2015 at 21:27:08, Spencer Oberstadt (notifications@github.com) wrote:

Could you give me some more information? How is the color getting set?

Currently, all styles are defined by css. As this is a WebView rendering all of the content.

**Comment by soberstadt** [comment]
It actually seems like this is something that might be better executed in only javascript (or coffeescript). 

If you wanted to use the bash output, you will need to parse it when it comes into the update function like this: https://github.com/soberstadt/simple-clock-widget/blob/master/simple-clock.widget/index.coffee#L33-L45

But I think a more efficent way of dealing with this (and I'd like to move the simple clock widget to this) is in the afterRender function (https://github.com/felixhageloh/uebersicht#afterrender--domel). You can use a `setTimeout` in Javascript where you get the current time and then update the dom to reflect that.

Does that make sense?
(I say all of this even though I made the simple-clock widget and it uses bash to built the time of day. If you implement this the javascript way, I might copy some of the code into simple-clock)

**Comment by Zorgodon** [comment]
Thanks, yes, that made sense! I've started programming the whole thing in js, but i'm confused as to how to implement my javascript in Ubersicht. I know that 

command: """
{my command}
"""
runs a bash script, but what statement do I use to run a javascript?

I get a result when I run the following in my browser:

var d = new Date();
var weekday = new Array(7);
weekday[0]=  "Sunday";
weekday[1] = "Monday";
weekday[2] = "Tuesday";
weekday[3] = "Wednesday";
weekday[4] = "Thursday";
weekday[5] = "Friday";
weekday[6] = "Saturday";

var n = weekday[d.getDay()];

document.write(n)

document.write(' ')


var monthNames = [
  "January", "February", "March",
  "April", "May", "June", "July",
  "August", "September", "October",
  "November", "December"
];

var date = new Date();
var day = date.getDate();
var monthIndex = date.getMonth();
var year = date.getFullYear();

console.log(day, monthNames[monthIndex], year);
document.write(day + ' ' + monthNames[monthIndex]);

In an html page. 

In my widget, I have command:"the script above" as well as:

refreshFrequency: 500

style: """
  top: 200px
  right: 0px
  color: #FFFFFF

  .output
    font-family: Roboto-Italic
    font-size: 60px
    font-weight: 100
    text-shadow: 0 0px 0px #000000;
"""

render: (output) -> """
  <div class="output">#{output}</div>
"""
at the bottom.

Thanks.
-- 
Ivan Bogush
Sent with Airmail

On 6 November 2015 at 14:17:32, Spencer Oberstadt (notifications@github.com) wrote:

It actually seems like this is something that might be better executed in only javascript (or coffeescript).

If you wanted to use the bash output, you will need to parse it when it comes into the update function like this: https://github.com/soberstadt/simple-clock-widget/blob/master/simple-clock.widget/index.coffee#L33-L45

But I think a more efficent way of dealing with this (and I'd like to move the simple clock widget to this) is in the afterRender function (https://github.com/felixhageloh/uebersicht#afterrender--domel). You can use a setTimeout in Javascript where you get the current time and then update the dom to reflect that.

Does that make sense?
(I say all of this even though I made the simple-clock widget and it uses bash to built the time of day. If you implement this the javascript way, I might copy some of the code into simple-clock)

**Comment by felixhageloh** [comment]
not 100% sure if this will answer (a part) of your question: 
You can skip running a bash command altogether by doing something like this

```coffeescript
command: ''

refreshFrequency: 1000

days: [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
]

render: ->
  now = new Date()
  h = now.getHours()
  m = now.getMinutes()
  s = now.getSeconds()
  
  # put any other js code here. make sure to return 
  # a string tho
  "<div>#{@days[now.getDay()]} <b>#{h}:#{m}:#{s}</b></div>"
```

Basically what is happening is that you just run an empty command (not the most efficient and something that could be improved in the future), but your widget still updates every second. On every update, the `render` method gets called, so you can put whatever JavaScript/CoffeeScript code you want to execute there.

**Comment by soberstadt** [comment]
Yeah, I would even go as far as setting the frequency to 0:
```
refreshFrequency: 0

afterRender: (domEl) ->
  var days = [
    'Sunday',
    'Monday',
    'Tuesday',
    'Wednesday',
    'Thursday',
    'Friday',
    'Saturday'
  ]
  function updateTime() {
    now = new Date()
    d = now.getDay()
    h = now.getHours()
    m = now.getMinutes()
    s = now.getSeconds()

    $(domEl).html("<div>"+days[d]+" <b>"+h+":"+m+":"+s+"</b></div>")
  }
  // run the update at the beginning
  updateTime()
  // run it every 1000 milliseconds
  setTimeout(updateTime, 1000)
```

**Comment by Zorgodon** [comment]
Thanks guys. Meanwhile i've done what I wanted to achieve using just a very simple bash command, with lots of javascript and css formatting (inspired by the simple date extension). Attached is my result. However, I do look forward to working with pure js and creating more fun widgets + stuff without using bash commands as you just explained. 

-- 
Ivan Bogush
Sent with Airmail

On 10 November 2015 at 19:53:49, Felix (notifications@github.com) wrote:

not 100% sure if this will answer (a part) of your question:
You can skip running a bash command altogether by doing something like this

command: ''

refreshFrequency: 1000

days: [
  'Sunday',
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday'
]

render: ->
  now = new Date()
  h = now.getHours()
  m = now.getMinutes()
  s = now.getSeconds()

  # put any other js code here. make sure to return  
  # a string tho
  "<div>#{@days[now.getDay()]} <b>#{h}:#{m}:#{s}</b></div>"
Basically what is happening is that you just run an empty command (not the most efficient and something that could be improved in the future), but your widget still updates every second. On every update, the render method gets called, so you can put whatever JavaScript/CoffeeScript code you want to execute there.
